src/compiler/binder.ts

```diff
@@ -482,6 +482,16 @@ module ts {
                         break;
                     }
                 case SyntaxKind.Block:
+                    // do not treat function block a block-scope container
+                    // all block-scope locals that reside in this block should go to the function locals.
+                    // Otherwise this won't be considered as redeclaration of a block scoped local:
+                    // function foo() {
+                    //  let x;
+                    //  var x;
+                    // }
+                    // 'var x' will be placed into the function locals and 'let x' - into the locals of the block

```

Comment:
Yeah, i'd add a comment to that effect.  I think using hte function as the location makes sense though, as that's where we put parameters.